src/util/taproot.rs

```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        &self.0
+    }
+
+    /// Create a merkle proof from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, TaprootError> {
+        if sl.len() % TAPROOT_CONTROL_NODE_SIZE != 0 {
+            Err(TaprootError::InvalidMerkleBranchSize(sl.len()))
+        } else if sl.len() > TAPROOT_CONTROL_NODE_SIZE * TAPROOT_CONTROL_MAX_NODE_COUNT {
+            Err(TaprootError::InvalidMerkleTreeDepth(
+                sl.len() / TAPROOT_CONTROL_NODE_SIZE,
+            ))
+        } else {
+            let inner = sl
+                // TODO: Use chunks_exact after MSRV changes to 1.31
+                .chunks(TAPROOT_CONTROL_NODE_SIZE)
+                .map(|chunk| {
+                    sha256::Hash::from_slice(chunk)
+                        .expect("chunk exact always returns the correct size")
+                })
+                .collect();
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
+
+    /// Serialize to a writer. Returns the number of bytes written
+    pub fn serialize_to_writer<Write: io::Write>(&self, mut writer: Write) -> io::Result<usize> {
+        let mut written = 0;
+        for hash in self.0.iter() {
+            written += writer.write(hash)?;
+        }
+        Ok(written)
+    }
+
+    /// Serialize self as bytes
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(TAPROOT_CONTROL_NODE_SIZE * self.0.len());
+        self.serialize_to_writer(&mut buf)
+            .expect("Already allocated buffer");
+        buf
+    }
+
+    // Internal function to append elements to proof
+    fn push(&mut self, h: sha256::Hash) -> Result<(), TaprootBuilderError> {
+        if self.0.len() + 1 > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len()))
+        } else {
+            self.0.push(h);
+            Ok(())
+        }
+    }
+
+    /// Create a MerkleProof from Vec<[`sha256::Hash`]>. Returns an error when
+    /// inner proof len is more than TAPROOT_CONTROL_MAX_NODE_COUNT (128)
+    pub fn from_inner(inner: Vec<sha256::Hash>) -> Result<Self, TaprootError> {
+        if inner.len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            Err(TaprootError::InvalidMerkleTreeDepth(inner.len()))
+        } else {
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
+
+    /// Consume Self to get Vec<[`sha256::Hash`]>
+    pub fn into_inner(self) -> Vec<sha256::Hash> {
+        self.0
+    }
+}
+
+/// Control Block data structure used in Tapscript satisfaction
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ControlBlock {
+    /// The internal key
+    pub internal_key: schnorr::PublicKey,
+    /// The parity of the output key (NOT THE INTERNAL KEY WHICH IS ALWAYS XONLY)
+    pub output_key_parity: bool,
+    /// The tapleaf version,
+    pub leaf_version: LeafVersion,
+    /// The merkle proof of a script associated with this leaf
+    pub merkle_branch: TaprootMerkleBranch,
```

Comment:
I would prefer to have the same field ordering as in serialization: leaf version, parity, internal key and merkle path